File: src/main/java/com/soccerbots/control/network/NetworkManager.java
package com.soccerbots.control.network;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.*;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class NetworkManager {
    private static final Logger logger = LoggerFactory.getLogger(NetworkManager.class);

    // ESP32 Communication Constants
    public static final int DISCOVERY_PORT = 12345;
    public static final int COMMAND_PORT_BASE = 12346;
    public static final int ESP32_UDP_PORT = 2367; // Legacy port
    public static final String EXPECTED_WIFI_NETWORK = "WATCHTOWER";

    private boolean isConnectedToNetwork = false;
    private String currentSSID = "";
    private ExecutorService executorService;
    private DatagramSocket udpSocket;
    private DatagramSocket discoverySocket;

    public NetworkManager() {
        this.executorService = Executors.newCachedThreadPool();
        checkCurrentNetworkStatus();
        initializeUDPSocket();
        initializeDiscoverySocket();
    }

    private void initializeUDPSocket() {
        try {
            udpSocket = new DatagramSocket();
            logger.info("UDP socket initialized for ESP32 communication");
        } catch (SocketException e) {
            logger.error("Failed to initialize UDP socket", e);
        }
    }

    private void initializeDiscoverySocket() {
        try {
            discoverySocket = new DatagramSocket(DISCOVERY_PORT);
            discoverySocket.setSoTimeout(100); // 100ms timeout for non-blocking receives
            logger.info("Discovery socket initialized on port {}", DISCOVERY_PORT);
        } catch (SocketException e) {
            logger.error("Failed to initialize discovery socket", e);
        }
    }

    private void checkCurrentNetworkStatus() {
        try {
            Process process = Runtime.getRuntime().exec("netsh wlan show interfaces");
            java.io.BufferedReader reader = new java.io.BufferedReader(
                new java.io.InputStreamReader(process.getInputStream()));

            String line;
            boolean connected = false;
            String ssid = "";

            while ((line = reader.readLine()) != null) {
                if (line.contains("State") && line.contains("connected")) {
                    connected = true;
                } else if (line.contains("SSID") && !line.contains("BSSID")) {
                    ssid = line.split(":")[1].trim();
                }
            }

            isConnectedToNetwork = connected;
            currentSSID = ssid;

            process.waitFor();

            if (connected) {
                logger.info("Connected to WiFi network: {}", ssid);
                if (EXPECTED_WIFI_NETWORK.equals(ssid)) {
                    logger.info("Connected to expected ESP32 network!");
                } else {
                    logger.warn("Connected to network '{}', but ESP32 robots expect '{}'",
                              ssid, EXPECTED_WIFI_NETWORK);
                }
            } else {
                logger.warn("Not connected to any WiFi network");
            }

        } catch (Exception e) {
            logger.error("Failed to check network status", e);
        }
    }

    /**
     * Send binary command data to ESP32 robot
     * Format: robotName(16 bytes) + axes(6 bytes) + buttons(2 bytes)
     */
    public void sendRobotCommand(String robotName, String targetIP,
                                int leftX, int leftY, int rightX, int rightY,
                                boolean cross, boolean circle, boolean square, boolean triangle) {
        executorService.submit(() -> {
            try {
                // Create 24-byte packet: 16 bytes name + 6 bytes axes + 2 bytes buttons
                byte[] packet = new byte[24];

                // Robot name (16 bytes, null-padded)
                byte[] nameBytes = robotName.getBytes();
                System.arraycopy(nameBytes, 0, packet, 0, Math.min(nameBytes.length, 16));

                // Axes data (6 bytes) - convert from float to 0-255 range
                packet[16] = (byte) Math.max(0, Math.min(255, leftX));   // leftX
                packet[17] = (byte) Math.max(0, Math.min(255, leftY));   // leftY
                packet[18] = (byte) Math.max(0, Math.min(255, rightX));  // rightX
                packet[19] = (byte) Math.max(0, Math.min(255, rightY));  // rightY
                packet[20] = (byte) 125; // unused axis
                packet[21] = (byte) 125; // unused axis

                // Button data (2 bytes)
                byte button1 = 0;
                if (cross) button1 |= 0x01;
                if (circle) button1 |= 0x02;
                if (square) button1 |= 0x04;
                if (triangle) button1 |= 0x08;

                packet[22] = button1;
                packet[23] = 0; // unused buttons

                // Send UDP packet
                DatagramPacket udpPacket = new DatagramPacket(
                    packet, packet.length,
                    InetAddress.getByName(targetIP), ESP32_UDP_PORT
                );
                udpSocket.send(udpPacket);

                logger.debug("Sent command to robot '{}' at {}:{}", robotName, targetIP, ESP32_UDP_PORT);

            } catch (Exception e) {
                logger.error("Failed to send robot command to " + targetIP, e);
            }
        });
    }

    /**
     * Send game status command to ESP32 robot
     * Format: "robotName:status" (text)
     */
    public void sendGameStatus(String robotName, String targetIP, String status) {
        executorService.submit(() -> {
            try {
                String message = robotName + ":" + status;
                byte[] data = message.getBytes();

                DatagramPacket packet = new DatagramPacket(
                    data, data.length,
                    InetAddress.getByName(targetIP), ESP32_UDP_PORT
                );
                udpSocket.send(packet);

                logger.info("Sent game status '{}' to robot '{}' at {}", status, robotName, targetIP);

            } catch (Exception e) {
                logger.error("Failed to send game status to " + targetIP, e);
            }
        });
    }

    /**
     * Broadcast game status to all robots on the network
     */
    public void broadcastGameStatus(String status) {
        try {
            // Get network info to determine broadcast address
            String subnet = getNetworkSubnet();
            if (subnet != null) {
                String broadcastAddr = subnet + ".255";
                logger.info("Broadcasting game status '{}' to {}", status, broadcastAddr);

                // Send to broadcast address - robots will filter by their name
                sendGameStatus("ALL", broadcastAddr, status);
            }
        } catch (Exception e) {
            logger.error("Failed to broadcast game status", e);
        }
    }

    private String getNetworkSubnet() {
        try {
            Process process = Runtime.getRuntime().exec("ipconfig");
            java.io.BufferedReader reader = new java.io.BufferedReader(
                new java.io.InputStreamReader(process.getInputStream()));

            String line;
            while ((line = reader.readLine()) != null) {
                if (line.contains("IPv4 Address") && line.contains("192.168")) {
                    String ip = line.split(":")[1].trim();
                    String[] parts = ip.split("\\.");
                    if (parts.length >= 3) {
                        return parts[0] + "." + parts[1] + "." + parts[2];
                    }
                }
            }
            process.waitFor();
        } catch (Exception e) {
            logger.error("Failed to get network subnet", e);
        }
        return null;
    }

    public boolean isNetworkActive() {
        checkCurrentNetworkStatus(); // Refresh status
        return isConnectedToNetwork;
    }

    public boolean isConnectedToExpectedNetwork() {
        return isConnectedToNetwork && EXPECTED_WIFI_NETWORK.equals(currentSSID);
    }

    public String getCurrentSSID() {
        return currentSSID;
    }

    public String getExpectedNetwork() {
        return EXPECTED_WIFI_NETWORK;
    }

    /**
     * Receive discovery message from robot
     * Returns null if no message available (non-blocking)
     */
    public String receiveDiscoveryMessage() {
        try {
            byte[] buffer = new byte[256];
            DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
            discoverySocket.receive(packet);
            String message = new String(packet.getData(), 0, packet.getLength());
            logger.debug("Received discovery message: {}", message);
            return message;
        } catch (SocketTimeoutException e) {
            // No message available, this is normal
            return null;
        } catch (Exception e) {
            logger.error("Error receiving discovery message", e);
            return null;
        }
    }

    /**
     * Send discovery response to robot
     */
    public void sendDiscoveryResponse(String targetIP, String message) {
        executorService.submit(() -> {
            try {
                byte[] data = message.getBytes();
                DatagramPacket packet = new DatagramPacket(
                    data, data.length,
                    InetAddress.getByName(targetIP), DISCOVERY_PORT
                );
                udpSocket.send(packet);
                logger.debug("Sent discovery response to {}: {}", targetIP, message);
            } catch (Exception e) {
                logger.error("Failed to send discovery response to " + targetIP, e);
            }
        });
    }

    /**
     * Send emergency stop to robot
     */
    public void sendEmergencyStop(String targetIP) {
        executorService.submit(() -> {
            try {
                String message = "ESTOP";
                byte[] data = message.getBytes();

                // Send to discovery port (robot always listens here when connected)
                DatagramPacket packet = new DatagramPacket(
                    data, data.length,
                    InetAddress.getByName(targetIP), DISCOVERY_PORT
                );
                udpSocket.send(packet);

                logger.info("Sent emergency stop to {}", targetIP);
            } catch (Exception e) {
                logger.error("Failed to send emergency stop to " + targetIP, e);
            }
        });
    }

    /**
     * Send emergency stop release to robot
     */
    public void sendEmergencyStopRelease(String targetIP) {
        executorService.submit(() -> {
            try {
                String message = "ESTOP_OFF";
                byte[] data = message.getBytes();

                DatagramPacket packet = new DatagramPacket(
                    data, data.length,
                    InetAddress.getByName(targetIP), DISCOVERY_PORT
                );
                udpSocket.send(packet);

                logger.info("Sent emergency stop release to {}", targetIP);
            } catch (Exception e) {
                logger.error("Failed to send emergency stop release to " + targetIP, e);
            }
        });
    }

    public void shutdown() {
        logger.info("Shutting down network manager");
        if (udpSocket != null && !udpSocket.isClosed()) {
            udpSocket.close();
        }
        if (discoverySocket != null && !discoverySocket.isClosed()) {
            discoverySocket.close();
        }
        if (executorService != null) {
            executorService.shutdown();
            try {
                if (!executorService.awaitTermination(2, java.util.concurrent.TimeUnit.SECONDS)) {
                    executorService.shutdownNow();
                }
            } catch (InterruptedException e) {
                executorService.shutdownNow();
                Thread.currentThread().interrupt();
            }
        }
    }
}

File: python_backend/network_manager.py
"""
Network Manager for ESP32 robot communication via UDP.
Handles discovery protocol and command transmission.
"""

import socket
import threading
import logging
import struct
import netifaces
from typing import Optional, Tuple

logger = logging.getLogger(__name__)


class NetworkManager:
    """Manages UDP communication with ESP32 robots"""
    
    # ESP32 Communication Constants
    DISCOVERY_PORT = 12345
    ESP32_UDP_PORT = 2367
    EXPECTED_WIFI_NETWORK = "WATCHTOWER"
    
    def __init__(self):
        self.udp_socket: Optional[socket.socket] = None
        self.discovery_socket: Optional[socket.socket] = None
        self.is_connected_to_network = False
        self.current_ssid = ""
        self._lock = threading.Lock()
        
        self._initialize_udp_socket()
        self._initialize_discovery_socket()
        self._check_current_network_status()
    
    def _initialize_udp_socket(self):
        """Initialize UDP socket for robot commands"""
        try:
            self.udp_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self.udp_socket.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
            logger.info("UDP socket initialized for ESP32 communication")
        except Exception as e:
            logger.error(f"Failed to initialize UDP socket: {e}")
    
    def _initialize_discovery_socket(self):
        """Initialize UDP socket for robot discovery"""
        try:
            self.discovery_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            self.discovery_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.discovery_socket.bind(('', self.DISCOVERY_PORT))
            self.discovery_socket.settimeout(0.1)  # 100ms timeout for non-blocking receives
            logger.info(f"Discovery socket initialized on port {self.DISCOVERY_PORT}")
        except Exception as e:
            logger.error(f"Failed to initialize discovery socket: {e}")
    
    def _check_current_network_status(self):
        """Check current WiFi network connection"""
        try:
            # Get active network interfaces
            interfaces = netifaces.interfaces()
            for interface in interfaces:
                addrs = netifaces.ifaddresses(interface)
                if netifaces.AF_INET in addrs:
                    for addr in addrs[netifaces.AF_INET]:
                        if 'addr' in addr and not addr['addr'].startswith('127.'):
                            self.is_connected_to_network = True
                            logger.info(f"Connected to network on interface {interface}: {addr['addr']}")
                            return
            
            logger.warn("Not connected to any network")
        except Exception as e:
            logger.error(f"Failed to check network status: {e}")
    
    def send_robot_command(self, robot_name: str, target_ip: str,
                          left_x: int, left_y: int, right_x: int, right_y: int,
                          cross: bool = False, circle: bool = False,
                          square: bool = False, triangle: bool = False):
        """
        Send binary command data to ESP32 robot.
        Format: robotName(16 bytes) + axes(6 bytes) + buttons(2 bytes) = 24 bytes total
        """
        if not self.udp_socket:
            logger.error("UDP socket not initialized")
            return
        
        try:
            # Create 24-byte packet
            packet = bytearray(24)
            
            # Robot name (16 bytes, null-padded)
            name_bytes = robot_name.encode('utf-8')[:16]
            packet[0:len(name_bytes)] = name_bytes
            
            # Axes data (6 bytes) - values 0-255
            packet[16] = max(0, min(255, left_x))
            packet[17] = max(0, min(255, left_y))
            packet[18] = max(0, min(255, right_x))
            packet[19] = max(0, min(255, right_y))
            packet[20] = 125  # unused axis
            packet[21] = 125  # unused axis
            
            # Button data (2 bytes)
            button1 = 0
            if cross: button1 |= 0x01
            if circle: button1 |= 0x02
            if square: button1 |= 0x04
            if triangle: button1 |= 0x08
            
            packet[22] = button1
            packet[23] = 0  # unused buttons
            
            # Send UDP packet
            self.udp_socket.sendto(bytes(packet), (target_ip, self.ESP32_UDP_PORT))
            logger.debug(f"Sent command to robot '{robot_name}' at {target_ip}:{self.ESP32_UDP_PORT}")
            
        except Exception as e:
            logger.error(f"Failed to send robot command to {target_ip}: {e}")
    
    def send_game_status(self, robot_name: str, target_ip: str, status: str):
        """
        Send game status command to ESP32 robot.
        Format: "robotName:status" (text)
        """
        if not self.udp_socket:
            logger.error("UDP socket not initialized")
            return
        
        try:
            message = f"{robot_name}:{status}"
            data = message.encode('utf-8')
            
            self.udp_socket.sendto(data, (target_ip, self.ESP32_UDP_PORT))
            logger.info(f"Sent game status '{status}' to robot '{robot_name}' at {target_ip}")
            
        except Exception as e:
            logger.error(f"Failed to send game status to {target_ip}: {e}")
    
    def broadcast_emergency_stop(self, activate: bool = True):
        """Broadcast emergency stop to all robots on discovery port"""
        if not self.udp_socket:
            logger.error("UDP socket not initialized")
            return
        
        try:
            message = "ESTOP" if activate else "ESTOP_OFF"
            data = message.encode('utf-8')
            
            # Send to broadcast address on discovery port (always monitored by robots)
            self.udp_socket.sendto(data, ('<broadcast>', self.DISCOVERY_PORT))
            logger.info(f"Broadcast emergency stop: {message}")
            
        except Exception as e:
            logger.error(f"Failed to broadcast emergency stop: {e}")
    
    def receive_discovery_message(self) -> Optional[str]:
        """Receive discovery message from robots (non-blocking)"""
        if not self.discovery_socket:
            return None
        
        try:
            data, addr = self.discovery_socket.recvfrom(1024)
            message = data.decode('utf-8')
            return message
        except socket.timeout:
            return None
        except Exception as e:
            logger.debug(f"Discovery receive error: {e}")
            return None
    
    def get_network_subnet(self) -> Optional[str]:
        """Get current network subnet (e.g., '192.168.1')"""
        try:
            interfaces = netifaces.interfaces()
            for interface in interfaces:
                addrs = netifaces.ifaddresses(interface)
                if netifaces.AF_INET in addrs:
                    for addr in addrs[netifaces.AF_INET]:
                        if 'addr' in addr and addr['addr'].startswith('192.168'):
                            parts = addr['addr'].split('.')
                            if len(parts) >= 3:
                                return f"{parts[0]}.{parts[1]}.{parts[2]}"
            return None
        except Exception as e:
            logger.error(f"Failed to get network subnet: {e}")
            return None
    
    def shutdown(self):
        """Cleanup network resources"""
        if self.udp_socket:
            self.udp_socket.close()
        if self.discovery_socket:
            self.discovery_socket.close()
        logger.info("Network manager shutdown complete")
